src/Api/Controller/ShowDiscussionController.php

```diff
@@ -77,7 +84,11 @@ protected function data(ServerRequestInterface $request, Document $document)
         $actor = $request->getAttribute('actor');
         $include = $this->extractInclude($request);
 
-        $discussion = $this->discussions->findOrFail($discussionId, $actor);
+        if (Arr::get($request->getQueryParams(), 'bySlug', false)) {
+            $discussion = $this->slugManager->forResource(Discussion::class)->fromSlug($discussionId, $actor);
+        } else {
+            $discussion = $this->discussions->findOrFail($discussionId, $actor);
+        }
```

Comment:
Looking at this I expect that we do this on every controller. I'd rather see (proposal):

- Repository::fromRequest method
- Repository loading the slug manager (this could even just be protected $slug = DiscussionSlug::class in the repository)
- Interface in Repository that implies it has sluggable resources to enforce specific methods?

This would reduce:

- Size of the controller and its bindings
- Increase size of the repository and the (late) bindings required; please note that the arguments in the controller methods are always instantiated soon, delegating this to the repository grants us a minor performance gain as well.
- Cleans up the conditional here, although it is moved elsewhere, we might even abstract the method `slugOrId` to an abstract repository or trait (SluggableRepository)

Let me know what you think.

Reply:
Mixed feelings about this. Currently, slug drivers use the repository. With this setup, the repository injects a slug manager which injects slug drivers. This feels convoluted to me. More importantly though, I feel like Flarum's internal use of repositories could really use some refinement: it's undocumented, and each repository seems to be assembled very haphazardly. That being said, I don't think this PR is the right place for that.